### PR TITLE
Inital idt setup

### PR DESCRIPTION
**Pull request by Jlisowskyy**
Things done in this PR:
- added stub array of isrs (Interrupt Service Routines),
- initialized IDT (Interrupt description table for 64bit code),
- added temporary code to print numeric values,
- added todo.hpp header,
- added initial asserts,
- added testing verifying if 0 division exception is correctly caught

Example execution log:
```
[INFO]	  Initializing Alkos
[OK]	  Detected Multiboot support
[OK]	  Detected CPUID support
[OK]	  Detected long mode support
[OK]	  Setup page tables
[OK]	  Enabled long mode
[OK]	  Enabled paging
[INFO]	  Jumping to 64-bit mode
[OK]	  In 64-bit mode
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 17 Starting pre-kernel initialization...
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 20 Starting to setup CPU features...
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 21 Setting up OS XSAVE...
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 23 Setting up SSE...
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 25 Setting up AVX...
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 27 Finished cpu features setup.
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 29 Setting up IDT...
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 31 IDT setup done.
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/init.cpp 33 Pre-kernel initialization finished.
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/kernel.cpp 23 Running kernel initialization...
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/src/kernel.cpp 28 Running test: 3
[INFO] /home/Jlisowskyy/Repos/AlkOS/alkos/kernel/arch/x86_64/interrupts/idt.cpp 70 Exception caught...
[ KERNEL PANIC ]
Unknown Exception caught -> default handler invoked.
```

### CODE REVIEW

File: alkos/kernel/arch/x86_64/interrupts/idt.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+#include <panic.hpp>
+
+TODO_BY_THE_END_OF_MILESTONE0
+/**
+ * @todo Add various debug traces when snprintf available
+ */
+
+/* crucial defines */
+static constexpr u32 kIdtEntries  = 256;
+static constexpr u8 kDefaultFlags = 0x8E;
```
Comment (kryczkal) [inline alkos/kernel/arch/x86_64/interrupts/idt.cpp:16]: maybe describe default flags used here in comment

File: alkos/kernel/arch/x86_64/linker.ld [hunk 2]
```diff
@@ -22,17 +22,17 @@ SECTIONS
...
+        KEEP(*(.multiboot))
 	}
 
 	. = ALIGN(4K);
 
 	.text32 BLOCK(4K) : ALIGN(4K)
-          {
-                    *(.text32)
-          }
+    {
```
Comment (kryczkal) [inline alkos/kernel/arch/x86_64/linker.ld:33]: i believe misaligning this is a mistake, but ehh not really that important

File: alkos/kernel/include/init.hpp [hunk 3]
```diff
@@ -3,4 +3,9 @@
 
 void KernelInit();
 
+/**
+ * @brief must be defined by arch code to set up all architecture dependent interrupt settings
+ */
+void KernelArchInit();
```
Comment (kryczkal) [inline alkos/kernel/include/init.hpp:9]: This is something to talk about.
I believe we can simplify the code by having just 2 setup functions for the arch
one being something like PreKernelInit32, other being like PreKernelInit64.
A PreKernelInit64 would setup things like IDT along with other things that we already set up. This way we would lessen the amount of functions that are required to be provided by the Arch as an ABI. Which translates to less places where we can fuck something up.
TLDR
I think this should be in arch during boot sequence and not here.
unless for some reason architecture specific things need to be setup after boot32, boot64 AND kernel

File: alkos/kernel/include/kernel_assert.hpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#define ASSERT(expr)                                                                                                   \
+    if (!(expr))                                                                                                       \
+    {                                                                                                                  \
+        FAIL_KERNEL(expr);                                                                                             \
+    }
+
+#endif // NDEBUG
+
+/* release assert */
+#define ASSERT_ALWAYS(expr)                                                                                            \
```
Comment (kryczkal) [inline alkos/kernel/include/kernel_assert.hpp:31]: I don't understand:
1. The purpose of this (The code for this is literally the same as normal ASSERT)
2. The semantics behind this. Assert always? Tf that even means. I don't know what you want this to be or wanted to achieve.

File: alkos/kernel/include/todo.hpp [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#ifndef KERNEL_INCLUDE_TODO_HPP_
```
Comment (kryczkal) [inline alkos/kernel/include/todo.hpp:1]: I don't intuitively get what's the use case. Maybe explain this a bit using a real world scenario plz

File: alkos/kernel/include/types.hpp [hunk 6]
```diff
@@ -5,4 +5,20 @@
 #include <stddef.h>
```
Comment (kryczkal) [inline alkos/kernel/include/types.hpp:5]: I like this a lot

File: alkos/kernel/src/init.cpp [hunk 7]
```diff
@@ -5,36 +5,17 @@
...
-     * 2. GDT, IDT, TSS, etc.
-     */
-
     TerminalInit();
     TRACE_INFO("Starting pre-kernel initialization...");
-    __stack_chk_init();
 
-    /* NOTE: sequence is important */
     TRACE_INFO("Starting to setup CPU features...");
-    TRACE_INFO("Setting up OS XSAVE...");
```
Comment (kryczkal) [inline alkos/kernel/src/init.cpp:28]: Again, i believe this should be in boot64, called before kernel.
The kernel should be run with maximum amount of arch dependent things already done. It also stinks to me that arch dependend setups are called inside the kernel.

The terminal is a bit of a different case - since there is no other way of implementation than providing an abi for drivers,  and it needs be used in later places of the code. but for thing like init where it's done once and forgot - its a different case.

File: alkos/kernel/src/panic.cpp [hunk 8]
```diff
@@ -1,13 +1,12 @@
...
 #include <panic.hpp>
 #include <terminal.hpp>
 
-extern "C" void KernelPanic(const char *msg) {
+extern "C" void KernelPanic(const char *msg)
+{
     TerminalWriteError("[ KERNEL PANIC ]\n");
     TerminalWriteError(msg);
 
-    asm volatile("cli");
```
Comment (kryczkal) [inline alkos/kernel/src/panic.cpp:9]: Another thing that needs to be provided by the abi. Inavoidable, yes - that's why i think we need to keep track of those things, like have a list of  functions needed to be provided by the arch in some file or list. And keep it minimal

**Review by kryczkal — APPROVED** [review]

**Comment by kryczkal** [comment]
One thing i dislike is the usage of ArchInit inside kernel code, more in comment
Other things are minor issues, or stuff i literally just don't understand from the code like assert_always or todo.hpp

**Comment by kryczkal** [comment]
Another thing - currently the list of functions that the architecture code needs to implement is unspecified, and kind of "growing in the dark" - I think that should be made clear and documented somewhere, otherwise we risk it quickly becoming so entangled with the kernel code that the distinction between arch and kernel dirs will be neglible

File: alkos/kernel/arch/x86_64/interrupts/idt.cpp (hunk 1 above)
Comment (Jlisowskyy) [reply alkos/kernel/arch/x86_64/interrupts/idt.cpp:16]: Sure great idea. To be honest I forgot to do it at the first place

File: alkos/kernel/arch/x86_64/linker.ld (hunk 2 above)
Comment (Jlisowskyy) [reply alkos/kernel/arch/x86_64/linker.ld:33]: I think it was wrongly aligned before so I fixed it

File: alkos/kernel/include/kernel_assert.hpp (hunk 4 above)
Comment (Jlisowskyy) [reply alkos/kernel/include/kernel_assert.hpp:31]: Basic C assert works only in debug mode we also need something which works in release mode. Here I proposed ASSERT_ALWAYS which workws always no matter build type which is strongly correlated with usual assert. We can work on names for example ASSERT would work as this macro and ASSERT_ONLY_IN_DEBUG like C-style assert

File: alkos/kernel/include/todo.hpp (hunk 5 above)
Comment (Jlisowskyy) [reply alkos/kernel/include/todo.hpp:1]: This is long discussion. Will be addressed during some sync call.

File: alkos/kernel/include/types.hpp (hunk 6 above)
Comment (Jlisowskyy) [reply alkos/kernel/include/types.hpp:5]: cool af

File: alkos/kernel/src/init.cpp (hunk 7 above)
Comment (Jlisowskyy) [reply alkos/kernel/src/init.cpp:28]: Will be addressed on some sync call.

File: alkos/kernel/src/panic.cpp (hunk 8 above)
Comment (Jlisowskyy) [reply alkos/kernel/src/panic.cpp:9]: Will be addressed on some sync call.

File: alkos/kernel/include/init.hpp (hunk 3 above)
Comment (Jlisowskyy) [reply alkos/kernel/include/init.hpp:9]: move do arch code

File: alkos/kernel/include/kernel_assert.hpp (hunk 4 above)
Comment (Jlisowskyy) [reply alkos/kernel/include/kernel_assert.hpp:31]: done

File: alkos/kernel/arch/x86_64/interrupts/idt.cpp (hunk 1 above)
Comment (Jlisowskyy) [reply alkos/kernel/arch/x86_64/interrupts/idt.cpp:16]: done
